### PR TITLE
Fix unnest for array of rows

### PR DESCRIPTION
**Pull request by pramodsatya**
The Unnest operator unfolds arrays of rows into a single column containing the row elements, which is referred to as legacy unnest in Presto. Unnest is expected to unfold each row element into as many columns as the row's cardinality, which is currently not supported in Velox. The following error is therefore seen in Prestissimo, since four integer columns are expected in the output but the legacy implementation of unnest returns two columns of row elements:
```
select * from unnest(ARRAY[row(2, 3), row(3, 5)], ARRAY[row(20, 30), row(30, 50)]) AS r(r1, r2, r3, r4);:

Field not found: field_1. Available fields are: field, field_0.
```
This PR adds support to unnest arrays of rows completely and fixes the above error. Legacy unnest will not be supported in Prestissimo and a follow-up PR in Prestodb will add the check to fail fast in case legacy unnest is requested. Resolves https://github.com/prestodb/presto/issues/20643

### CODE REVIEW

File: velox/exec/Unnest.cpp [hunk 1]
```diff
@@ -192,13 +192,29 @@ RowVectorPtr Unnest::getOutput() {
...
-          : wrapChild(
-                numElements,
-                elementIndices,
-                unnestBaseArray->elements(),
-                nulls);
+      auto elements = unnestBaseArray->elements();
+
+      if (elements->typeKind() == TypeKind::ROW) {
+        auto rowSize = elements->type()->size();
+        auto rowDecoded = DecodedVector(*elements);
```
Comment (mbasmanova) [inline velox/exec/Unnest.cpp:199]: DecodedVector rowDecoded(*elements);

File: velox/exec/Unnest.cpp [hunk 2]
```diff
@@ -192,13 +192,29 @@ RowVectorPtr Unnest::getOutput() {
...
+            rowVector,
+            "Unable to get rowVector from elements with {} encoding",
+            elements->encoding());
+
+        for (auto j = 0; j < rowSize; j++) {
+          VELOX_CHECK(rowVector->childAt(j), "Column at index {} is null", j);
+          outputs[outputsIndex++] = identityMapping
+              ? rowVector->childAt(j)
+              : wrapChild(
+                    numElements, elementIndices, rowVector->childAt(j), nulls);
```
Comment (mbasmanova) [inline velox/exec/Unnest.cpp:211]: This is incorrect as it ignores encodings on top of rowVector.

File: velox/exec/tests/UnnestTest.cpp [hunk 3]
```diff
@@ -37,6 +37,31 @@ TEST_F(UnnestTest, basicArray) {
 
   auto op = PlanBuilder().values({vector}).unnest({"c0"}, {"c1"}).planNode();
   assertQuery(op, "SELECT c0, UNNEST(c1) FROM tmp WHERE c0 % 7 > 0");
+
+  // Test with array containing dictionary encoded elements.
+  vector_size_t size = 200;
+  auto offsets = makeIndices(size, [](auto row) { return row; });
```
Comment (mbasmanova) [inline velox/exec/tests/UnnestTest.cpp:43]: Try with non-identify mapping. For example, makeIndicesInReverse.

**Comment by aditi-pandit** [comment]
@pramodsatya : There were bunch more changes to get this right. We should close this in favor of https://github.com/facebookincubator/velox/pull/7477

**Pull request closed without merging**